src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs

```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+        }
+        var parent = codeNamespace.FindChildByName<CodeFunction>($"deserializeInto{returnType}");
+
+        return conventions.GetTypeString(new CodeType { TypeDefinition = parent }, codeElement, false);
+    }
+
+    private void WriteSerializerFunction(CodeFunction codeElement, LanguageWriter writer)
+    {
+        var param = codeElement.OriginalLocalMethod.Parameters.FirstOrDefault(x => ((CodeType)x.Type).TypeDefinition is CodeInterface);
+        if (param == null || param.Type is not CodeType codeType || codeType.TypeDefinition == null)
+
+            throw new InvalidOperationException("Interface parameter not found for code interface");
+
+        var codeInterface = (CodeInterface)codeType.TypeDefinition;
+        var inherits = codeInterface.StartBlock.Implements.FirstOrDefault(x => x.TypeDefinition is CodeInterface);
+        writer.IncreaseIndent();
+
+        if (inherits != null)
+        {
+            writer.WriteLine($"serialize{inherits?.TypeDefinition?.Name.ToFirstCharacterUpperCase()}(writer, {param.Name.ToFirstCharacterLowerCase()})");
+        }
+
+        foreach (var otherProp in codeInterface.Properties.Where(static x => x.Kind == CodePropertyKind.Custom && !x.ExistsInBaseType && !x.ReadOnly))
+        {
+            WritePropertySerializer(codeInterface.Name.ToFirstCharacterLowerCase(), otherProp, writer, codeElement);
+        }
+
+        var additionalDataProperty = codeInterface.GetPropertyOfKind(CodePropertyKind.AdditionalData);
+        if (additionalDataProperty != null)
+            writer.WriteLine($"writer.writeAdditionalData({codeInterface.Name.ToFirstCharacterLowerCase()}.{additionalDataProperty.Name.ToFirstCharacterLowerCase()});");
+        writer.DecreaseIndent();
+    }
+
+    private static bool IsCodePropertyCollectionOfEnum(CodeProperty property)
+    {
+        return property.Type is CodeType cType && cType.IsCollection && cType.TypeDefinition is CodeEnum;
+    }
+
+    private void WritePropertySerializer(string modelParamName, CodeProperty codeProperty, LanguageWriter writer, CodeFunction codeFunction)
+    {
+        var isCollectionOfEnum = IsCodePropertyCollectionOfEnum(codeProperty);
+        var spreadOperator = isCollectionOfEnum ? "..." : string.Empty;
+        var codePropertyName = codeProperty.Name.ToFirstCharacterLowerCase();
+
+        var propertyTypeName = (codeProperty.Type as CodeType)?.TypeDefinition?.Name;
+        if (codeProperty.Parent == null)
+        {
+            throw new InvalidOperationException($"Code property must have parent {codePropertyName}");
+        }
+        var propType = localConventions?.GetTypeString(codeProperty.Type, codeProperty.Parent, false);
+
+
+        var serializationName = GetSerializationMethodName(codeProperty.Type);
+
+        if (serializationName == "writeObjectValue" || serializationName == "writeCollectionOfObjectValues")
```

Comment:
move to a static readonly hashset with an ordinal ignore case comparer and use the contains method instead